### PR TITLE
don't use FlexiBLAS on POWER for foss toolchain, use OpenBLAS instead

### PR DESCRIPTION
**Pull request by branfosj**
(created using `eb --new-pr`)

Explanation for suggesting this change is in https://github.com/easybuilders/easybuild-easyconfigs/pull/13836

### CODE REVIEW

File: easybuild/toolchains/foss.py [hunk 1]
```diff
@@ -56,7 +57,7 @@ def __init__(self, *args, **kwargs):
         constants = ('BLAS_MODULE_NAME', 'BLAS_LIB', 'BLAS_LIB_MT', 'BLAS_FAMILY',
                      'LAPACK_MODULE_NAME', 'LAPACK_IS_BLAS', 'LAPACK_FAMILY')
 
-        if self.looseversion > LooseVersion('2021.0'):
+        if self.looseversion > LooseVersion('2021.0') and get_cpu_architecture() != POWER:
```
Comment (boegel) [inline easybuild/toolchains/foss.py:60]: @branfosj We should explain a bit better why we opt out on POWER, please add a comment incl. a reference to https://github.com/easybuilders/easybuild-easyconfigs/issues/12968 and https://github.com/mpimd-csc/flexiblas/issues/17 ?

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/toolchains/foss.py (hunk 1 above)
Comment (branfosj) [reply easybuild/toolchains/foss.py:63]: comment added in 8ab4aef

**Comment by boegel** [comment]
@branfosj As we've discussed, I'm hesitant to bake this workaround into an EasyBuild release, since rolling this back later when we have a proper fix (a patch for GCC to fix the miscompilation issue) will cause more pain for sites using `foss/2021a` on POWER systems: they'll need to reinstall their `foss/2021a` module and all modules that were built with the `foss/2021a` toolchain, to ensure that everything is linking consistently to FlexiBLAS.

I spent some time in trying to find a `-fno-*` compiler option that disables an optimization pass that causes the miscompilation in FlexiBLAS or OpenBLAS (it's not 100% clear to me where it goes wrong exactly, I guess it's actually in OpenBLAS but the problem is only exposed when using it via FlexiBLAS?), based on the info mentioned in the [GCC bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100799) that the miscompilation doesn't occur with `-Og`, but failed to find one...

I'll tag this for the release after EasyBuild v4.4.2 to keep it in view, but we should pursue other ways to fix this issue..

For sites who want to use `foss/2021a` on POWER:

* You can include this modified `foss.py` into your active EasyBuild installation via [`--include-toolchains`](https://docs.easybuild.io/en/latest/Including_additional_Python_modules.html#including-additional-toolchains-include-toolchains), there's no need to manually patch your EasyBuild installation.
* You can use the modified `foss-2021a.eb` easyconfig from https://github.com/easybuilders/easybuild-easyconfigs/pull/13836, and ensure that EasyBuild picks up on that (rather than the one included with EasyBuild) via [`--robot-paths`](https://docs.easybuild.io/en/latest/Using_the_EasyBuild_command_line.html#controlling-robot-search-path).

cc @Flamefire @verdurin

**Comment by Flamefire** [comment]
>  guess it's actually in OpenBLAS but the problem is only exposed when using it via FlexiBLAS?),

No the miscompilation happens in FlexiBLAS. There might be some at other places but that one is the one I found and which is causing issues. Might be due to the "trampoline function" used by flexiblas which just redirects to the real BLAS and that way somehow runs into an optimizer bug.

From my side I wouldn't mind using OpenBLAS for POWER with 2021a. If there will be a fix soonish then it will be for 2021b, but I kinda doubt there will

**Comment by akesandgren** [comment]
Any updates on the miscompilation part?

**Comment by branfosj** [comment]
I am no longer working on this, so I am closing this. The information in https://github.com/easybuilders/easybuild-framework/pull/3809#issuecomment-912968575 says how to setup an alternative foss toolchain on POWER to include these changes.

**Comment by boegel** [comment]
Proper workaround for allowing the use of FlexiBLAS on POWER is implemented in https://github.com/easybuilders/easybuild-easyconfigs/pull/15885, so there's no need anymore for special workarounds for POWER by mangling the definition of the `foss` toolchain...

**Pull request closed without merging**